### PR TITLE
[ibmcloud-appid] Add types

### PR DESCRIPTION
**Pull request by younes-io**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/ibmcloud-appid/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
+// Type definitions for ibmcloud-appid 6.2
+// Project: https://github.com/ibm-cloud-security/appid-serversdk-nodejs
+// Definitions by: Younes A <https://github.com/younes-io>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/*~ If this module is a UMD module that exposes a global variable 'myLib' when
```
Comment (peterblazejewicz) [inline types/ibmcloud-appid/index.d.ts:6]:
```suggestion
/*~ If this module is a UMD module that exposes a global variable 'myLib' when
```
please remove this block and any other block which are obvious instructional commands from template. Add your own or leave without comments

Comment (younes-io) [reply types/ibmcloud-appid/index.d.ts]: Done

**Comment by younes-io** [comment]
The merge of this PR will depend on this [PR](https://github.com/ibm-cloud-security/appid-serversdk-nodejs/pull/268)

File: types/ibmcloud-appid/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+// tslint:disable-next-line:no-unnecessary-class
+export class SelfServiceManager {
+    constructor(options: SelfServiceOptions);
+    signUp: (userData: unknown, language: string, iamToken: string) => Promise<UserSCIM>;
+}
+
+// tslint:disable-next-line:no-unnecessary-class
+export class UserProfileManager {
+    constructor();
```
Comment (peterblazejewicz) [inline types/ibmcloud-appid/index.d.ts:63]:
```suggestion
```

File: types/ibmcloud-appid/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+// tslint:disable-next-line:no-unnecessary-class
+export class UnauthorizedException {
+    constructor();
```
Comment (peterblazejewicz) [inline types/ibmcloud-appid/index.d.ts:68]:
```suggestion
```

File: types/ibmcloud-appid/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+export class TokenManager {
+    constructor(options: StrategyOptions);
+    getApplicationIdentityToken: () => Promise<ApplicationIdentityToken | Error>;
+    getCustomIdentityTokens: () => Promise<CustomIdentityToken | Error>;
+}
+
+// tslint:disable-next-line:no-unnecessary-class
+export class SelfServiceManager {
+    constructor(options: SelfServiceOptions);
+    signUp: (userData: unknown, language: string, iamToken: string) => Promise<UserSCIM>;
```
Comment (peterblazejewicz) [inline types/ibmcloud-appid/index.d.ts:58]:
```suggestion
    signUp: (userData: { [key: string]: unknown }, language: string, iamToken: string) => Promise<UserSCIM>;
```
> Sign up a new user. userData is a JSON object

so use something silar to this keyed object definition, not pain `unknown`

File: types/ibmcloud-appid/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+export interface SelfServiceOptions {
+    iamApiKey?: string;
+    managementUrl?: string;
+    tenantId?: string;
+    oAuthServerUrl?: string;
+    iamTokenUrl?: string;
+}
+
+export class Strategy {
```
Comment (peterblazejewicz) [inline types/ibmcloud-appid/index.d.ts:18]: you cannot export this as a symmbol, original module does not do this AFAIK:
https://github.com/ibm-cloud-security/appid-serversdk-nodejs/blob/master/lib/appid-sdk.js#L30-L37

File: types/ibmcloud-appid/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+// tslint:disable-next-line:no-unnecessary-class
+export class TokenManager {
+    constructor(options: StrategyOptions);
+    getApplicationIdentityToken: () => Promise<ApplicationIdentityToken | Error>;
+    getCustomIdentityTokens: () => Promise<CustomIdentityToken | Error>;
+}
+
+// tslint:disable-next-line:no-unnecessary-class
```
Comment (peterblazejewicz) [inline types/ibmcloud-appid/index.d.ts:55]:
```suggestion
```

File: types/ibmcloud-appid/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+export class APIStrategy extends Strategy {
+    constructor(options: StrategyOptions);
+}
+
+// tslint:disable-next-line:no-unnecessary-class
+export class WebAppStrategy extends Strategy {
+    constructor(options: StrategyOptions);
+}
+
+// tslint:disable-next-line:no-unnecessary-class
```
Comment (peterblazejewicz) [inline types/ibmcloud-appid/index.d.ts:48]:
```suggestion
```

**Comment by peterblazejewicz** [comment]
please also review, if you could support exact optional property types in your options, we are promoting this usage to provide betteer libraries interop (e..g when some user set this option for his project):
https://www.typescriptlang.org/play?ts=4.4.2&exactOptionalPropertyTypes=true#example/exact-optional-properties

**Review by sandersn — CHANGES_REQUESTED** [review]
Requesting changes so the bot takes @peterblazejewicz 's review seriously.

**Pull request closed without merging**